0000-classical-stores.md

```diff
@@ -0,0 +1,493 @@
+In particular, this RFC does not yet address the question of making `Expr` nodes usable in `params` locations (such as the angle inputs to gates).
+
+
+## Motivation and Benefit
+
+Qiskit needs a way to represent the sorts of "parametric update" that we expect hardware vendors to be implementing in the near future.
+For example, we will ideally be able to use a single compiled circuit (both in the Qiskit and hardware senses of the term "compiled") that accepts certain parameters as inputs at runtime, which may affect conditional expressions or gate angles.
+This RFC only addresses the question of defining new classical-variable storage locations and writing to them for Qiskit within its existing typed `Expr` system, which is currently limited to unsigned integers and Booleans and cannot be used for gate angles.
+
+For example, Pauli-twirling workloads want to do something like
+```python
+from qiskit.circuit import QuantumCircuit
+from qiskit.circuit.classical import expr
+
+qc = QuantumCircuit(2)
+with qc.if_test(expr.value(TWIRL_BITS_0)):
+    qc.x(0)
+with qc.if_test(expr.value(TWIRL_BITS_1)):
+    qc.x(1)
+```
+where `TWIRL_BITS_*` are some values defined by the particular set of twirls to be done, and are known before the circuit begins execution, but not _necessarily_ at the time of the definition of the circuit, or the circuit should be run with many different values for them.
+These may also want to be compared to the results of mid-circuit measurements, or have arbitrary run-time classical expressions built from them.
+Right now, Qiskit provides no way to parametrise `if` conditions in any form; even the compile-time rewriting of `assign_parameters` will not support them, because `ParameterExpression` objects are (deliberately) not valid in `condition` fields.
+
+Similarly, a user may need to calculate a value based on some measured bits, potentially dependent on some `if` condition, and store the value in classical memory for later use in the circuit.
+In other words, they need variable assignment at runtime.
+
+
+## Design Proposal
+
+At a high level, this proposal concerns adding concepts equivalent to the OpenQASM 3 `input` (and potentially `output`) and variable-assignment-by-expression statements to Qiskit, within the classical types we already support.
+I do not yet propose removing `Clbit` or `ClassicalRegister`, nor trying to change them into some other form; for the time being, they retain their "special-case" status.
+
+The additional user interface for defining and working with variables on `QuantumCircuit` is:
+
+```python
+class QuantumCircuit:
+    def __init__(
+        self,
+        ...,  # existing structure
+        *,
+        inputs: Iterable[expr.Var] = (),
+        captures: Iterable[expr.Var] = (),
+        declarations: Mapping[expr.Var, expr.Expr] | Iterable[Tuple[expr.Var, expr.Expr]] = (),
+    ): ...
+
+    def add_var(self, name_or_var: str | expr.Var, / initial: expr.Expr) -> expr.Var: ...
```

Comment:
I think you are missing a `type: types.Type` between `name_or_var` and `/`.
```suggestion
    def add_var(self, name_or_var: str | expr.Var, type: types.Type, / initial: expr.Expr) -> expr.Var: ...
```

Reply:
No, it's deliberately not there because it's always inferable, so having it would only really offer an opportunity for desynchronisation and errors.

Reply:
In all your examples below you provide a type, though, e.g.
```
qc.add_var("mask", types.Uint(3), expr.value(5))
```

Reply:
Fixed in 1dc9c4569f9d9274246d2baae98eb2961cf467bb.